File: front/src/libs/service/taskHandsOnService.ts
import axios from 'axios';

const API_URL = process.env.NEXT_PUBLIC_API_URL;
const BASE_URL = API_URL ? `${API_URL}/api/task_hands_on` : undefined;

export type HandsOnGenerationConfig = Record<string, unknown>;

export interface HandsOnGenerationRequest {
  project_id: string;
  config?: HandsOnGenerationConfig;
}

export interface HandsOnGenerationResponse {
  success: boolean;
  job_id: string;
  project_id: string;
  status: string;
  total_tasks: number;
  message: string;
}

export interface HandsOnJobStatusProgress {
  total_tasks: number;
  completed: number;
  failed: number;
  processing: number;
  pending: number;
}

export interface HandsOnProcessingTask {
  task_id: string;
  task_title: string;
}

export interface HandsOnCompletedTask {
  task_id: string;
  task_title: string;
  quality_score: number | null;
  completed_at: string | null;
}

export interface HandsOnJobStatusResponse {
  success: boolean;
  job_id: string;
  project_id: string;
  status: string;
  progress: HandsOnJobStatusProgress;
  current_processing: HandsOnProcessingTask[];
  completed_tasks: HandsOnCompletedTask[];
  error_message?: string | null;
  error_details?: Record<string, unknown> | null;
  created_at?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
}

export interface HandsOnTargetFile {
  path: string;
  description: string;
  action: string;
}

export interface HandsOnCodeExample {
  file: string;
  language: string;
  code: string;
  explanation?: string;
}

export interface HandsOnCommonError {
  error: string;
  cause: string;
  solution: string;
}

export interface HandsOnImplementationTip {
  type: 'best_practice' | 'anti_pattern';
  tip: string;
  reason: string;
}

export interface HandsOnReference {
  title: string;
  url: string;
  type?: string;
}

export interface HandsOnContent {
  hands_on_id: string;
  overview: string | null;
  prerequisites: string | null;
  target_files: HandsOnTargetFile[] | null;
  implementation_steps: string | null;
  code_examples: HandsOnCodeExample[] | null;
  verification: string | null;
  common_errors: HandsOnCommonError[] | null;
  references: HandsOnReference[] | null;
  technical_context: string | null;
  implementation_tips: HandsOnImplementationTip[] | null;
}

export interface HandsOnMetadata {
  generated_at: string | null;
  quality_score: number | null;
  generation_model: string | null;
  information_freshness: string | null;
  search_queries: string[] | null;
  referenced_urls: string[] | null;
}

export interface TaskHandsOnResponse<THandsOn = Record<string, unknown>, TMetadata = Record<string, unknown>> {
  success: boolean;
  task_id: string;
  task_title: string;
  has_hands_on: boolean;
  hands_on?: THandsOn | null;
  metadata?: TMetadata | null;
}

export interface DeleteHandsOnResponse {
  success: boolean;
  deleted_count: number;
  message: string;
}

export interface HandsOnPreviewResponse {
  success: boolean;
  preview_mode?: boolean;
  message: string;
  [key: string]: unknown;
}

const ensureBaseUrl = (): string => {
  if (!BASE_URL) {
    throw new Error('NEXT_PUBLIC_API_URL is not defined');
  }
  return BASE_URL;
};

export const startHandsOnGeneration = async (
  request: HandsOnGenerationRequest,
): Promise<HandsOnGenerationResponse> => {
  const baseUrl = ensureBaseUrl();
  const response = await axios.post<HandsOnGenerationResponse>(`${baseUrl}/generate_all`, request);
  return response.data;
};

export const fetchHandsOnJobStatus = async (
  jobId: string,
): Promise<HandsOnJobStatusResponse> => {
  const baseUrl = ensureBaseUrl();
  const response = await axios.get<HandsOnJobStatusResponse>(`${baseUrl}/status/${jobId}`);
  return response.data;
};

export const fetchTaskHandsOn = async <THandsOn = HandsOnContent, TMetadata = HandsOnMetadata>(
  taskId: string,
): Promise<TaskHandsOnResponse<THandsOn, TMetadata>> => {
  const baseUrl = ensureBaseUrl();
  const response = await axios.get<TaskHandsOnResponse<THandsOn, TMetadata>>(`${baseUrl}/${taskId}`);
  return response.data;
};

export const deleteProjectHandsOn = async (
  projectId: string,
): Promise<DeleteHandsOnResponse> => {
  const baseUrl = ensureBaseUrl();
  const response = await axios.delete<DeleteHandsOnResponse>(`${baseUrl}/${projectId}`);
  return response.data;
};

export const previewHandsOnGeneration = async (
  request: HandsOnGenerationRequest,
): Promise<HandsOnPreviewResponse> => {
  const baseUrl = ensureBaseUrl();
  const response = await axios.post<HandsOnPreviewResponse>(`${baseUrl}/preview`, request);
  return response.data;
};

File: back/services/task_hands_on_service.py
"""
TaskHandsOnService: タスクハンズオン管理サービス

インタラクティブモード移行後の簡略化版
- 一括生成機能は廃止（interactive_hands_on_agent.py に移行）
- CRUD操作のみをサポート
"""

from typing import Dict, Optional
from sqlalchemy.orm import Session
from uuid import UUID

from models.project_base import (
    Task,
    TaskHandsOn,
)


class TaskHandsOnService:
    """
    タスクハンズオン管理サービス（CRUD操作のみ）

    インタラクティブ生成は interactive_hands_on_agent.py で行う
    """

    def __init__(self, db: Session):
        self.db = db

    def get_task_hands_on(self, task_id: UUID) -> Optional[Dict]:
        """
        タスクのハンズオンを取得

        Args:
            task_id: タスクID

        Returns:
            ハンズオン辞書（存在しない場合はhas_hands_on=False）
        """
        task = self.db.query(Task).filter_by(task_id=task_id).first()
        if not task:
            raise ValueError(f"Task {task_id} not found")

        hands_on = self.db.query(TaskHandsOn).filter_by(task_id=task_id).first()

        if not hands_on:
            return {
                "task_id": str(task.task_id),
                "task_title": task.title,
                "has_hands_on": False,
                "hands_on": None,
                "metadata": None
            }

        return {
            "task_id": str(task.task_id),
            "task_title": task.title,
            "has_hands_on": True,
            "hands_on": {
                "hands_on_id": str(hands_on.hands_on_id),
                "overview": hands_on.overview,
                "prerequisites": hands_on.prerequisites,
                "target_files": hands_on.target_files,
                "implementation_steps": hands_on.implementation_steps,
                "code_examples": hands_on.code_examples,
                "verification": hands_on.verification,
                "common_errors": hands_on.common_errors,
                "references": hands_on.references,
                "technical_context": hands_on.technical_context,
                "implementation_tips": hands_on.implementation_tips,
            },
            "metadata": {
                "generated_at": hands_on.generated_at.isoformat() if hands_on.generated_at else None,
                "quality_score": hands_on.quality_score,
                "generation_model": hands_on.generation_model,
                "information_freshness": hands_on.information_freshness.isoformat() if hands_on.information_freshness else None,
                "search_queries": hands_on.search_queries,
                "referenced_urls": hands_on.referenced_urls,
            }
        }

    def delete_project_hands_on(self, project_id: UUID) -> int:
        """
        プロジェクトの全ハンズオンを削除

        Args:
            project_id: プロジェクトID

        Returns:
            削除件数
        """
        # プロジェクトのタスクIDを取得
        task_ids = [
            task.task_id
            for task in self.db.query(Task).filter_by(project_id=project_id).all()
        ]

        if not task_ids:
            return 0

        # ハンズオンを削除
        deleted_count = (
            self.db.query(TaskHandsOn)
            .filter(TaskHandsOn.task_id.in_(task_ids))
            .delete(synchronize_session=False)
        )

        self.db.commit()

        return deleted_count
